src/magic.h

```diff
@@ -385,6 +385,15 @@ class spell_type
         static void check_consistency();
         static void reset_all();
         bool is_valid() const;
+
+        // these two formulas should be the inverse of eachother.  The spell xp will break if this is not the case.
+        std::optional<jmath_func_id> get_level_formula_id;
+        std::optional<jmath_func_id> exp_for_level_formula_id;
```

Comment:
I admit I don't really like the idea of doing a loop to find the level.  It also feels like it may be a performance downgrade since players could potentially be calculating for the level 30-40 times per spell which is the multiplied by every spell visible and then even more since my prints counted get_level being called several times per spell per opening the spellbook.

In terms of difficulty, I don't feel that it would be an enormous burden to do the inversion with the current setup.  Both Magiclysm and XE currently use EOCs to mimic a constant xp / level for their spells, so the formulas they would swap to would be y=ax and x=y/a.  For potential other uses it would be more difficult, but since those would already require doing some math for 1 spell formula I don' think that contributor would have any significant issues inverting the formula that they calculate.